### ISSUE TITLE
Magento 2.3.1: Catalog setup fails with error "Magento\Catalog\Setup\Media does not exist"

### ISSUE THREAD

**Issue by nussjustin** [magento/magento2#22124]
### Preconditions
1. magento/product-enterprise-edition 2.3.0
2. PHP 7.2

Exact PHP version:

```
PHP 7.2.16 (cli) (built: Mar  5 2019 13:10:50) ( NTS )
Copyright (c) 1997-2018 The PHP Group
Zend Engine v3.2.0, Copyright (c) 1998-2018 Zend Technologies
    with Zend OPcache v7.2.16, Copyright (c) 1999-2018, by Zend Technologies
    with Xdebug v2.7.0, Copyright (c) 2002-2019, by Derick Rethans
```

### Steps to reproduce
1. Update from magento/product-enterprise-edition 2.3.0 to 2.3.1 
2. Run `bin/magento setup:upgrade`

### Expected result
1. No errors, normal output

### Actual result
1. Error `Class Magento\Catalog\Setup\Media does not exist` when running `Data install/update` for `Magento_Catalog`

This seems to be a bug in the `CategorySetup.php`:

https://github.com/magento/magento2/blob/4a28abb/app/code/Magento/Catalog/Setup/CategorySetup.php#L629

This should probably be `\Magento\Catalog\Model\Product\Attribute\Backend\Media::class` or a `use \Magento\Catalog\Model\Product\Attribute\Backend\Media` be added at the beginning of the file.

**Comment by nussjustin** [comment]
> Hi @nussjustin. Thank you for your report.
> To help us process this issue please make sure that you provided the following information:
> 
>   - [x] Summary of the issue
>   - [x] Information on your environment
>   - [x] Steps to reproduce
>   - [x] Expected and actual results
>
> Please make sure that the issue is reproducible on the vanilla Magento instance following Steps to reproduce. To deploy vanilla Magento instance on our environment, please, add a comment to the issue:
> 
> For more details, please, review the Magento Contributor Assistant documentation.
> 
> @nussjustin do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
> 
> - [x] yes
> - [ ] no

**Comment by hostep** [comment]
Nice find @nussjustin! Issue seems to have been introduced in https://github.com/magento/magento2/commit/dcd8b323950b924db07ba7b37898a9c3a8b0f2fd#diff-7b8ac6c93b8d79968c8d4d8237621404L587 (I'm wondering why the tests didn't catch this...)

Would you be willing to create a Pull Request with a fix for this?

**Comment by nussjustin** [comment]
After looking at it again, I'm not sure what the correct fix would be. In Magento 2.2 there was some kind of mapping in the di.xml

https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Catalog/etc/di.xml#L692-L703

This was removed in 2.3.1 at it seems, so just using `Magento\Catalog\Model\Product\Attribute\Backend\Media` does not work anymore.

**Comment by renttek** [comment]
maybe the author of the mentioned commit has an answer for this, @rganin ?

**Comment by orlangur** [comment]
This functionality does not seem to be a part of Magento Open Source, isn't it?

In such case
> Please report Magento Commerce issues via the Support portal of your account or Partner portal if you are a partner reporting on behalf of a merchant.
GitHub is intended for Magento Open Source bug reports given no account management for Magento Open Source users. This will allow for proper tracking of issues at the account level.

**Comment by renttek** [comment]
@orlangur he may have used the enterprise, but the error is in Magento_Catalog, which is in Open Source...

**Comment by orlangur** [comment]
@renttek it does not matter, it's quite a frequent situation that enterprise issues require fixes in community modules.

This issue can be reopened if it is reproducible on a Magento Open Source. Otherwise it just cannot be reproduced and fixed by Community members.

**Comment by renttek** [comment]
I can understand this reasoning, but this is literally a basic 'could be catched by an IDE' php coding error, in a an OS module, introduced by @rganin withint commit dcd8b323950b924db07ba7b37898a9c3a8b0f2fd (Which references "MAGETWO-87551")

@nussjustin maybe you should just open another issue with the CE m(

**Comment by orlangur** [comment]
@renttek man, do you hear me? :) There is no need to open another one.

@nussjustin @hostep @renttek could please anyone confirm it is reproducible on Magento Open Source? I really doubt this wasn't tested prior to release but reopening in advance.

**Comment by hostep** [comment]
@nussjustin: thanks for pointing out that `Magento\Catalog\Model\Product\Attribute\Backend\Media` actually doesn't even exist, hadn't seen that yet. It was removed a while ago in https://github.com/magento/magento2/commit/0f0063045b5 (part of merge commit 9b08b0b2eede6d6c6c70c9fc1485e8ca91f3fe16)
Regarding the mapping, I don't think this is relevant here.

If I take a look at the database, in the table `eav_attribute`, the column `backend_model` for the attribute `media_gallery` is set to `NULL`, both in Magento 2.3.1 as 2.2.7.
So it looks like providing a non existing class in the database migration scripts for the backend model doesn't throw errors or warnings, but just inserts `NULL`. (Not sure if that is correct though, a warning would be nice here I think?).

The correct fix here would probably be to remove the backend model in the `CategorySetup.php` file for the `media_gallery` attribute as it doesn't need to have one I think.

@orlangur: I don't know how to reproduce the error @nussjustin sees. But this is a code smell, and it would be nice if this could get cleaned up I think.

**Comment by orlangur** [comment]
> The correct fix here would probably be to remove the backend model in the CategorySetup.php file for the media_gallery attribute as it doesn't need to have one I think.

Agree on that, @nussjustin for the reproduction would be nice if you figure out what causes a problem in one of Magento Commerce modules, maybe provide an isolated case as a small module just for revealing issue.

**Comment by nussjustin** [comment]
I'm currently unable to reproduce with a fresh Magento 2 (with sample data, both when installing 2.3.1 directly and when upgrading from 2.3.0 to 2.3.1) both CE and EE, but consistently with a copy of one of our customers M2 EE installations.

This problem is currently preventing us from upgrading one of our customers sites from 2.3.0 EE to 2.3.1 EE.

**Comment by orlangur** [comment]
> If I take a look at the database, in the table eav_attribute, the column backend_model for the attribute media_gallery is set to NULL, both in Magento 2.3.1 as 2.2.7.

@nussjustin probably this is not true for this one customer?

As a workaround just create a dummy `app/code/Magento/Catalog/Model/Product/Attribute/Backend/Media.php` class and surely it would be nice to create a pull request for `2.3-develop` which eliminates referring nonexistent class.

**Comment by nussjustin** [comment]
@orlangur The backend_model for media_gallery is NULL for this customer.

Creating an empty class Media that extends AbstractBackend worked, thanks. The backend_model is still NULL in the eav_attribute table, but everything works.

**Comment by magento-engcom-team** [comment]
Hi @nussjustin. Thank you for your report. 
The issue has been fixed in magento/magento2#22446 by @hostep in 2.3-develop branch
Related commit(s):
 - [f22e16b6f7bd84303cfd22dd5dc17d9a58b8101b](https://github.com/magento/magento2/commit/f22e16b6f7bd84303cfd22dd5dc17d9a58b8101b)
 - [9e00ed29c722e0f77ea7e3dc98f7917dc2afdb9d](https://github.com/magento/magento2/commit/9e00ed29c722e0f77ea7e3dc98f7917dc2afdb9d)

The fix will be available with the upcoming 2.3.2 release.
